File: LeetCode/Valid Palindrome/ValidPalindrome.cs
// https://leetcode.com/problems/valid-palindrome/
// Details - Runtime: 92 ms, faster than 99.42% of C# online submissions for Valid Palindrome.

public class Solution {
    public bool IsPalindrome(string s) {
        int left = 0;
        int right = s.Length - 1;

        while (left < right)
        {
            while (!char.IsLetterOrDigit(s[left]) && (left < right))
            {
                ++left;
            }
            while (!char.IsLetterOrDigit(s[right]) && (right > left))
            {
                --right;
            }
            if (char.ToLower(s[left]) != char.ToLower(s[right]))
            {
                return false;
            }
            ++left;
            --right;
        }
        return true;
    }
}

File: LeetCode/Valid Palindrome/ValidPalindrome.java
// https://leetcode.com/problems/valid-palindrome/submissions/
// Details - Runtime: 6 ms, faster than 95.30% of Java online submissions for Valid Palindrome.

class Solution {
    public boolean isPalindrome(String s) {
        char[] chars = s.toCharArray();
        int left = 0;
        int right = s.length() - 1;

        while (left < right)
        {
            while (!Character.isLetterOrDigit(chars[left]) && (left < right))
            {
                ++left;
            }
            while (!Character.isLetterOrDigit(chars[right]) && (right > left))
            {
                --right;
            }
            if (Character.toLowerCase(chars[left]) != Character.toLowerCase(chars[right]))
            {
                return false;
            }
            ++left;
            --right;
        }
        return true;
    }
}
